Add `torch.nested` namespace

mikaylagawarecki opened the pull request:
First step towards #83775
- only `to_padded_tensor` is moved to the nested namespace for now
- following the schema used for `special`, `fft`, `linalg` and other namespaces, nested functions are registered in native_functions.yaml as `nested_{function_name}` and are bound to the desired Python name in
`torch/nested/__init__.py`, and the desired C++ name in `torch/csrc/api/include/torch/nested.h`.

~~**Question**: should we keep the documentation for `Tensor.to_padded_tensor` or can this deleted since it is shared by `torch.nested.to_padded_tensor`?~~

[generated nested docs](https://docs-preview.pytorch.org/84102/nested.html?highlight=nested#module-torch.nested)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #84102

Differential Revision: [D39361148](https://our.internmc.facebook.com/intern/diff/D39361148)

mikaylagawarecki committed:
Add `torch.nested` namespace

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add `torch.nested` namespace"




[ghstack-poisoned]

drisspg commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -12609,6 +12609,17 @@
+#   torch/nested/__init__.py, and the desired C++ name in torch/csrc/api/include/torch/nested.h.
+#   The "nested_" names should be hidden from the user and not documented.
+
+- func: nested_to_padded_tensor(Tensor self, float padding, int[]? output_size=None) -> Tensor
```
Is it the pattern to create aliases for namespaced functions? I think it makes more sense to only have one function. If the pattern is to create aliases for namespace functions then then we should create a name neseted_aliases.cpp and move the definitions there.

drisspg commented on aten/src/ATen/native/nested/NestedTensorMath.cpp:
```diff
@@ -414,6 +414,14 @@ Tensor NestedTensor_to_padded_tensor_generic(
   return ret_val;
 }
 
+// alias for to_padded_tensor in nested namespace
```
Going to go and refactor the nestedtensor math file and split to smaller more logical cpp files. Hence the comment above

mikaylagawarecki replied:
Added NestedTensorAliases.cpp, will import and include in the relevant bzl files

drisspg commented on caffe2/CMakeLists.txt:
```diff
@@ -398,6 +398,7 @@ if(NOT INTERN_BUILD_MOBILE OR NOT BUILD_CAFFE2_MOBILE)
     "${TORCH_SRC_DIR}/csrc/autograd/generated/python_nn_functions.cpp"
     "${TORCH_SRC_DIR}/csrc/autograd/generated/python_fft_functions.cpp"
     "${TORCH_SRC_DIR}/csrc/autograd/generated/python_linalg_functions.cpp"
+    "${TORCH_SRC_DIR}/csrc/autograd/generated/python_nested_functions.cpp"
```
Does this have internal impacts that need to be replicated in bzl files?

mikaylagawarecki replied:
Good catch, I probably need to import this and make those additions.

drisspg commented on docs/source/nested.rst:
```diff
@@ -142,14 +142,11 @@ nested_tensor([
 +++++++++++++++++++++++++
 
-The following Tensor methods are related to nested tensors:
+The following functions are related to nested tensors:
```
"The following functions are specific to nested tensors:" ?

mikaylagawarecki replied:
Hm, so these will be functions in the nested namespace, hence the change, but they still exist as methods. What do you think would be the best way of putting this across?

albanD replied:
I think that only functions should be here and methods will be documented on the Tensor object directly?

mikaylagawarecki replied:
Just to confirm, we still want to_padded_tensor to live as a function in the torch.nested namespace, is that correct?

drisspg replied:
Alban suggested to me that we could potentially remove the method on the tensor base. And use the top level python api to create the method. This python level implementation would under the hood call the torch.nested.to_padded_tensor() function

But yeah I think as long as there is one **function** exposed in python not having both `torch.to_padded_tensor(nt, 0)` and `torch.nested.to_padded_tensor(nt, 0 )` that is good.

drisspg commented on test/cpp/api/CMakeLists.txt:
```diff
@@ -18,6 +18,7 @@ set(TORCH_API_TEST_SOURCES
   ${TORCH_API_TEST_DIR}/moduledict.cpp
   ${TORCH_API_TEST_DIR}/modulelist.cpp
   ${TORCH_API_TEST_DIR}/modules.cpp
+  ${TORCH_API_TEST_DIR}/nested.cpp
```
same bazel again here

drisspg commented on test/test_nestedtensor.py:
```diff
@@ -274,6 +274,13 @@ def test_to_padded_tensor_on_empty_tensor(self):
+
+    def test_nested_namespace(self):
+        nt = torch.nested_tensor([torch.randn(2, 3), torch.randn(4, 5)])
+        result = nt.to_padded_tensor(4)
```
yeah I hope it isn't the pattern duplicate torch.* namespace functions into the aliases because having both feels really wrong

albanD replied:
This one is method-only. So there is nothing in `torch.*`.

drisspg replied:
Ohh see when then added a function that goes in nested namespace? Do we need that? Should only functional variants go in nested namespace?

albanD replied:
In general yes because you don't want nested-specific methods on plain Tensor.
But I guess it's too late :p

drisspg commented on torch/csrc/jit/passes/normalize_ops.cpp:
```diff
@@ -158,7 +159,8 @@ const std::unordered_map<Symbol, Symbol>& getOperatorAliasMap() {
       {aten::special_gammaincc, aten::igammac},
-      {aten::special_gammaln, aten::lgamma}};
+      {aten::special_gammaln, aten::lgamma},
+      {aten::nested_to_padded_tensor, aten::to_padded_tensor}};
```
I guess we do the aliasing thing..

albanD commented on test/allowlist_for_publicAPI.json:
```diff
@@ -1225,6 +1225,9 @@
   "torch.multiprocessing.spawn": [
     "Optional"
   ],
+  "torch.nested": [
```
Why do you need this since you already have a `__all__` there?

mikaylagawarecki replied:
TestCorrectModuleNames seemed to fail with just the `__all__`  before adding it here. I'll check again to make sure

mikaylagawarecki replied:
Double checked this, test_public_bindings.py:test_correct_module_names seems to fail without this

albanD commented on torch/_tensor_docs.py:
```diff
@@ -6511,55 +6511,3 @@ def callable(a, b) -> number
 """,
 )
-
-add_docstr_all(
```
You want to keep an entry for this method.
It can be simply `See torch.nested.to_padded_tensor()` (like many others in this file that point to torch.)

albanD reviewed (commented):
Sounds good!

mikaylagawarecki committed:
Update on "Add `torch.nested` namespace"

First step towards #83775
- only `to_padded_tensor` is moved to the nested namespace for now
- following the schema used for `special`, `fft`, `linalg` and other namespaces, nested functions are registered in native_functions.yaml as `nested_{function_name}` and are bound to the desired Python name in
`torch/nested/__init__.py`, and the desired C++ name in `torch/csrc/api/include/torch/nested.h`.

**Question**: should we keep the documentation for `Tensor.to_padded_tensor` or can this deleted since it is shared by `torch.nested.to_padded_tensor`?

[generated nested docs](https://docs-preview.pytorch.org/84102/nested.html?highlight=nested#module-torch.nested)




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add `torch.nested` namespace"

First step towards #83775
- only `to_padded_tensor` is moved to the nested namespace for now
- following the schema used for `special`, `fft`, `linalg` and other namespaces, nested functions are registered in native_functions.yaml as `nested_{function_name}` and are bound to the desired Python name in
`torch/nested/__init__.py`, and the desired C++ name in `torch/csrc/api/include/torch/nested.h`.

**Question**: should we keep the documentation for `Tensor.to_padded_tensor` or can this deleted since it is shared by `torch.nested.to_padded_tensor`?

[generated nested docs](https://docs-preview.pytorch.org/84102/nested.html?highlight=nested#module-torch.nested)




[ghstack-poisoned]

drisspg reviewed (commented):
One more comment, If the native function `to_padded_tensor` didn't exist and we were adding it to native functions.yaml would we only have this entry? : 
```
- func: nested_to_padded_tensor(Tensor self, float padding, int[]? output_size=None) -> Tensor
  python_module: nested
  variants: function
``` 

If so I think that we should consider removing [to_padded_tensor](https://github.com/pytorch/pytorch/blob/85931eaa6beab53d138f873d3505aee34e98ee89/aten/src/ATen/native/native_functions.yaml#L13073), which defines a method entirely. 

Is there a great reason to have both of these native_function entries, maybe I am missing something?

mikaylagawarecki commented:
@drisspg  I'm slightly concerned with breaking BC by removing the method since we had already unprivated and documented `Tensor.to_padded_tensor` :(

drisspg commented:
> @drisspg I'm slightly concerned with breaking BC by removing the method since we had already unprivated and documented `Tensor.to_padded_tensor` :(

Yeah but we aren't even prototype lol and no one really uses nested tensor. I think it would be okay. @albanD Thoughts?

Would an alternative if we still wanted to be a method in python be adding it as a method to this class:https://github.com/pytorch/pytorch/blob/65f98eb47dbf75335d08f7676835a5e1f1fc3574/torch/_tensor.py#L99

mikaylagawarecki committed:
Update on "Add `torch.nested` namespace"

First step towards #83775
- only `to_padded_tensor` is moved to the nested namespace for now
- following the schema used for `special`, `fft`, `linalg` and other namespaces, nested functions are registered in native_functions.yaml as `nested_{function_name}` and are bound to the desired Python name in
`torch/nested/__init__.py`, and the desired C++ name in `torch/csrc/api/include/torch/nested.h`.

~~**Question**: should we keep the documentation for `Tensor.to_padded_tensor` or can this deleted since it is shared by `torch.nested.to_padded_tensor`?~~

[generated nested docs](https://docs-preview.pytorch.org/84102/nested.html?highlight=nested#module-torch.nested)




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add `torch.nested` namespace"

First step towards #83775
- only `to_padded_tensor` is moved to the nested namespace for now
- following the schema used for `special`, `fft`, `linalg` and other namespaces, nested functions are registered in native_functions.yaml as `nested_{function_name}` and are bound to the desired Python name in
`torch/nested/__init__.py`, and the desired C++ name in `torch/csrc/api/include/torch/nested.h`.

~~**Question**: should we keep the documentation for `Tensor.to_padded_tensor` or can this deleted since it is shared by `torch.nested.to_padded_tensor`?~~

[generated nested docs](https://docs-preview.pytorch.org/84102/nested.html?highlight=nested#module-torch.nested)




[ghstack-poisoned]

mikaylagawarecki commented:
@pytorchbot rebase -s

mikaylagawarecki committed:
Update on "Add `torch.nested` namespace"

First step towards #83775
- only `to_padded_tensor` is moved to the nested namespace for now
- following the schema used for `special`, `fft`, `linalg` and other namespaces, nested functions are registered in native_functions.yaml as `nested_{function_name}` and are bound to the desired Python name in
`torch/nested/__init__.py`, and the desired C++ name in `torch/csrc/api/include/torch/nested.h`.

~~**Question**: should we keep the documentation for `Tensor.to_padded_tensor` or can this deleted since it is shared by `torch.nested.to_padded_tensor`?~~

[generated nested docs](https://docs-preview.pytorch.org/84102/nested.html?highlight=nested#module-torch.nested)




[ghstack-poisoned]

mikaylagawarecki commented:
git rebase -s

mikaylagawarecki commented:
@pytorchbot rebase -s

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39361148).

mikaylagawarecki commented:
Removed `Tensor.to_padded_tensor()` and imported + changed internal invocations of `Tensor.to_padded_tensor --> torch.nested.to_padded_tensor`

drisspg commented on build_variables.bzl:
```diff
@@ -983,6 +983,7 @@ def glob_libtorch_python_sources(gencode_pattern = ":generate-code[{}]"):
         "torch/csrc/autograd/generated/python_functions_2.cpp",
         "torch/csrc/autograd/generated/python_functions_3.cpp",
         "torch/csrc/autograd/generated/python_functions_4.cpp",
+    "torch/csrc/autograd/generated/python_nested_functions.cpp",
```
nit: is spacing weird on this?

mikaylagawarecki replied:
oh shoot thanks I didn't notice this

drisspg reviewed (approved):
Awesome great work Mikayla!

mikaylagawarecki committed:
Update on "Add `torch.nested` namespace"

First step towards #83775
- only `to_padded_tensor` is moved to the nested namespace for now
- following the schema used for `special`, `fft`, `linalg` and other namespaces, nested functions are registered in native_functions.yaml as `nested_{function_name}` and are bound to the desired Python name in
`torch/nested/__init__.py`, and the desired C++ name in `torch/csrc/api/include/torch/nested.h`.

~~**Question**: should we keep the documentation for `Tensor.to_padded_tensor` or can this deleted since it is shared by `torch.nested.to_padded_tensor`?~~

[generated nested docs](https://docs-preview.pytorch.org/84102/nested.html?highlight=nested#module-torch.nested)


Differential Revision: [D39361148](https://our.internmc.facebook.com/intern/diff/D39361148)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39361148).

mikaylagawarecki commented:
@pytorchbot rebase -s

mikaylagawarecki committed:
Update on "Add `torch.nested` namespace"

First step towards #83775
- only `to_padded_tensor` is moved to the nested namespace for now
- following the schema used for `special`, `fft`, `linalg` and other namespaces, nested functions are registered in native_functions.yaml as `nested_{function_name}` and are bound to the desired Python name in
`torch/nested/__init__.py`, and the desired C++ name in `torch/csrc/api/include/torch/nested.h`.

~~**Question**: should we keep the documentation for `Tensor.to_padded_tensor` or can this deleted since it is shared by `torch.nested.to_padded_tensor`?~~

[generated nested docs](https://docs-preview.pytorch.org/84102/nested.html?highlight=nested#module-torch.nested)


Differential Revision: [D39361148](https://our.internmc.facebook.com/intern/diff/D39361148)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39361148).

mikaylagawarecki commented:
@pytorchbot rebase -s

mikaylagawarecki commented:
@pytorchbot rebase -s

mikaylagawarecki committed:
Update on "Add `torch.nested` namespace"

First step towards #83775
- only `to_padded_tensor` is moved to the nested namespace for now
- following the schema used for `special`, `fft`, `linalg` and other namespaces, nested functions are registered in native_functions.yaml as `nested_{function_name}` and are bound to the desired Python name in
`torch/nested/__init__.py`, and the desired C++ name in `torch/csrc/api/include/torch/nested.h`.

~~**Question**: should we keep the documentation for `Tensor.to_padded_tensor` or can this deleted since it is shared by `torch.nested.to_padded_tensor`?~~

[generated nested docs](https://docs-preview.pytorch.org/84102/nested.html?highlight=nested#module-torch.nested)


Differential Revision: [D39361148](https://our.internmc.facebook.com/intern/diff/D39361148)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39361148).
